Support for CAGRA (type 14 GEB format) & (optionally) Grand Raiden at RCNP

csullivan opened the pull request:
This merge request brings in support for parsing type 14 GEB event data (Digital Gammasphere), which is useful for analyzing data from multiple systems, including the Clover Array Gamma-ray spectrometer at RCNP/RIBF -- CAGRA. It also brings in the necessary TDetector systems and event source for coupling GRUTinizer with the RCNP unpacker utilized with the Grand Raiden (GR) spectrometer and the Large Acceptance Spectrometer (LAS) (TRCNPSource library). This latter addition is optional and can be enabled by checking out the GRAnalyzer submodule from https://github.com/CAGRA-GrandRaiden/.

For the purpose of this merge request the GRAnalyzer submodule is not necessary and all of the supporting GRUTinizer libraries are only compiled if RCNP=1 is present in the makefile.

Please test with your normal workflow. I will address any issues if they rise.

Commit:
Adding new TDetector system for parsing
mode 3 data from the LBL digitizers with the Argonne
National Lab firmware. This will allow us to instrument
DGS and CAGRA data in GRUTinizer.

Commit:
Built out TArgonne(Hit) classes. The raw event
parser should now handle the Argonne National Lab data with firmware v1.1
and of the LED type. Support for other formats and versions will come later.
Added ANL as a detector system, and am now ready to adjust GRUTinizer
unpacking logic to handle events from raw data files.

Commit:
Added ANL kFileType and kDetectorSystem
and added TArgonne/ANL to the TUnpackingLoop,
and TUnpackedEvent case-switch statements to
build TArgonne event parer into GRUTinizer. I
have also added a check for the raw ANL event
file extension gtd* to TGRUTOptions. Finally,
I added endian swaps into the BuildHits.
Ready for testing.

Commit:
(incremental) kFileType::ANL -> kFileType::ANL_RAW

Commit:
(incremental) fixed enum miss match in TRawEventSource.

Commit:
ANL event building is now fully functional.
Added MakeANLhistos.cxx for analysis. chicka, chi-kaaah!

Commit:
(incremental) Set more members in TArgonneHit.

Commit:
Merge branch 'ANL'

Notes: Added support for the ANL firmware LBL digitizers,
including firmware v1.1 for Digital Gammasphere, and CAGRA.
Support for additional ANL firmware versions is upcoming.

Commit:
(incremental) added clover histograms

Commit:
Merge from upstream, updated ANL GEB classes to match new format. Mainly BuildHits is passed the raw_data reference
instead of TDetector storing internally

Commit:
Merge branch 'master' of https://github.com/pcbend/GRUTinizer

Commit:
Merge branch 'master' of https://github.com/pcbend/GRUTinizer

Commit:
Added RCNP Tree source class, needs implementation

Commit:
Added templated TTreeSource class for reading preprocessed events from a ROOT TTree

Commit:
Made default constructor of TTreeSource private

Commit:
Built out TTreeSource. Calling GetEvent effectively calls TTree::GetEntry, creating a NEW event in memory each time.
A pointer to this location in memory is serialized into the TSmartBuffer of the TRawEvent, which is passed on. GetEntry
fills fEvent which is the type over which TTreeSource is specialized. For now I added a prototype EventType class to demo
it's usage and how to register it in the LinkDef. This will be deleted after testing with data has been done.

Commit:
Added ability for ROOT file to be a source via TTreeSource. Added functionality to
TGRUTint::ApplyOptions and TRawEventSource::EventSource. Needs testing.

Commit:
Added necessary files for builiding TGrandRaiden data from a root file source.
Additionally, updated makefile to link against the external libRCNPTREE.so

Commit:
Removed trailing whitespace

Commit:
Successfully linking against GRAnalyzer data structure RCNPEvent.
Some issue exists in the root file not being picked up as a TTreeSource.

Commit:
Added TTreeSource<RCNPEvent> which now moves data through and into TGrandRaiden hits where unpacking can be done.
Time to test event correlation between CAGRA and GR!

Commit:
Added some members to TGrandRaidenHit to test the GR data stream

Commit:
Added fake timestamps to GR data to test event correlation. Bug fix in TRawEvent.

Commit:
Added histos for all CAGRA boards and channels

Commit:
Completely restructured Argonne/CAGRA interface. Instead of
having TArgonne and TArgonneHit, I refactored TArgonneHit into a TANLEvent type
that is responsible for doing the unpacking, and then the other classes are
solely responsible for building the unpacked event into calibrated events.
TArgonne/TArgonneHits have been renamed to TCAGRA/TCAGRAHits. Left to do:
Need figure out why Clover channel 0 is not appearing in histos. Need to
test to see if the positions map is working right. Need to test if the
calibrations file is working right. See if the segmentation is working or
if the TCAGRASegmentHit class is overkill. Finally, need expand channel
calibrations to include second order corrections such as asymptotic baseline
and pole-zero corrections, for example.

Commit:
Updated channel map file to include crystals _and_ segments. The positions now seem to be working.
Each clover detector gets a number from 01 - 16, and each clover has A,B,C,D leaf crystals. Each crystal can have segments or not.
The naming scheme works as follows: CLO01AP00 this indicates the system is CLO short for Clover. The next number 01 indicates this
is clover 01. A indicates the crystal number (should be the crystal leaf at smaller theta). P indicates that the signal is positive
and 00 means that this is the 0th segment (ie the central contact). Some detectors will have multiple segments in which case this
last number will be 01 or 02. The address is currently made from the BoardID and Channel number. So Board 113 (0x71) and channel 01
would be address 0x01007101. The first 01 is just for historical reasons, and indicates the subsystem number.

Commit:
Merge pull request #1 from CAGRA-GrandRaiden/RCNPSource

Merging RCNPSource branch maintained in CAGRA+GR organization

Commit:
Added calibrated energy spectra. The calibrations work!

Commit:
TCAGRA has been replaced by TCagra due to loss in pinky strength

Commit:
Bug introduced when refactoring into TANLEvent. Forgot to add check to
TCagra::BuildHits to distinguish between ArraySubpositions (Leaves of a clover). Now fixed.

Commit:
Merging RCNPSource into upstream master. This branch will replace csullivan/master

Commit:
List of changes:
Major: Fixed bug in TTreeSource::GetEvent in which the stack address was being copied into the TSmartBuffer, not the value at the stack address.
       This was what was causing a double free to occur.
       TGrandRaidenHit::BuiltHits and TTreeSource were adjusted to be more symmetric also
Major: TMultiRawFile changes are in progress. Adjusting so that it has slightly different functionality in online and offline modes.

Commit:
Merging from TMultiRawFile_StalledSource branch. Added online functionality to TMultiRawFile. See commit log.

Commit:
rootalyze.h -> RCNPEvent.h, also removed +1 to Timestamps, and I am now able to verify that all GR events are are in coincidence with cagra events, as expected with the fake timestamps

Commit:
Added RF to TGrandRaidenHit, also more histos. RCNPEvent->T in templated TTreeSource class. Added back print message in TDataLoop

Commit:
Added source description and status to TTreeSource, data throughput is not a measure of totale filesize in this case,
since the only value passed through is the ptr to the event created via the TChain::GetEntry call.

Commit:
Changed initialization of fEvent in TTreeSource

Commit:
Now link against new libGRAnalyzer.so for dynamic calling of GR event builder

Commit:
Added Grand Raiden analyzer submodule for linking against RCNP routines

Commit:
Removed commented lines in makefile

Commit:
Added lines to clean rule in makefile

Commit:
Updated submodule and granalyzer.cxx example

Commit:
Added support for RCNPEvent's in ThreadsafeQueue, and also adjusted the granalyzer example
so that a pipe command is passed to start_analyzer which can be used to cat a file in, or to launch the online data stream.

Commit:
Added new source for raw RCNP BLD files. GRUTinizer calls GRAnalyzer which prepares RCNPEvents
that are pushed on to a threadsafe queue and accessed by TRCNPSource. Added a TRCNPEvent which adds an RCNPEvent ptr to a
normal TRawEvent. Added needed logic for this new raw file type. Need to find a way to differential between file runs
and online runs with router_save, and to properly form pipe command for each scenario.

Commit:
Now spawn std::async thread which returns a future that can be used to test if the GRAnalyzer is done.
Still need a way to stop the GR thread while in progress if an early termination is desired. Added logic to Pop events off the
rcnp_queue in TRCNPSource. In offline mode the logic is set up to perform a second _saving throw_ pop if the first was unsuccessful.
If the second is also unsuccessful and if the GR ana thread is not finished, it throws an exception. In online mode it will
simply continue to ask for events and so this logic is not needed. But as mentioned above, we still need a flag to signal to the GR
analyzer thread to terminate if we are closing up early (this is especially true for online analysis).

Bug fix: There is an issue with the TSmartBuffer destructor that needs investigation. I am able to avoid a segfault
by declaring the smartbuffer in TTreeSource and TRCNPSource on the heap, and letting it leak out of scope. Needs further
investigation.

Commit:
Added global atomic_int that is set by StoppableThread StopAllClean function, which tells the RCNP
analyzer loop to terminate so that all threads can cleanly exit.

Commit:
Added global atomic_int that is set by StoppableThread StopAllClean function, which tells the RCNP
analyzer loop to terminate so that all threads can cleanly exit.

Commit:
Added -R option to indicate if the raw RCNP root tree should be saved while analyzing RCNP data.

Commit:
Added setup script to create the necessary links to the GRAnalyzer git submodule

Commit:
Changed rcnp gitmodule address

Commit:
Added line to source thisgrut.sh

Commit:
Isolated RCNP code a bit so that if you have the correct submodule (GRAnalyzer),
it can be enabled by setting RCNP=1 in the makefile, but otherwise, the
dependandt classes (TTreeSource TRCNPSource RCNPEvent etc) are implemented
as "husk" versions of themselves.  All defguards are in RCNP classes and
are not present in GRUTinizer for the sake of clean code.

Commit:
Removed unnecessary script

Commit:
Merge remote branch 'origin/RCNPUnified' into RCNPUnified

Commit:
Temporarily changing the submodule path for work at RCNP

Commit:
Added missing rcnpchannel calibration file to analyzer

Commit:
Merge branch 'RCNPUnified' of nsclgw1.nscl.msu.edu:~/ceclub/sullivan/cagragr/GRUTinizer into RCNPUnified

Commit:
Timestamps bath was absolute, and added <functional> to TCaeser.h
which is needed in 4.9.3

Commit:
Adding simple program for reading ANL GEB data

Commit:
Merge branch 'RCNPUnified' of nsclgw1.nscl.msu.edu:~/ceclub/sullivan/cagragr/GRUTinizer into RCNPUnified

Commit:
Changed gitmodules url for submodule to GitHub ssh

Commit:
Added functionality for ANL firmware v1.8 (June 2015). July 2015 additions will be necessary.

Commit:
Major: Added void* fDataPtr to TRawEvent. This allows for TRawEvent to hold user information other than what is in
the TSmartBuffer, which the user is responsible for cleaning up in one of their TDetector systems.

Commit:
Merge branch 'bug_GrandRaiden' into RCNPUnified

Commit:
Now RCNPEvent is carried through to the rootfile in TGrandRaidenHit.
No copies are done, since the copy constructor of RCNPEvent performs a swap on it's underlying data map.

Updated histograms loop to include GR histosgrams as discussed with Guillaume on 6/19. Since the RCNPEvent is saved
with each hit, the output root tree can be sorted without rerunning the analyzer, which is useful for calibration
steps.

Commit:
Improved makefile to check the hist.def file and also added RCNP=1 to makefile. Also added runtime checkt that the hist.def file has not changed from when it was first compiled

Commit:
Added MPIEventLoop submodule for parallel analysis. Also added fake timestamps back, added histo cut on gr_x.

Commit:
Fixed error in makefile, and added onlin.bld check to TRawEventSource::EventSource to start RCNP ana in online mode

Commit:
Implemented new CFDv18. Untested

Commit:
Fixed compile time issues with name mixups between LEDv18 and CFDv18.

Commit:
CFDv1.8 seems to be working, but need more rigourous checks, mainly with data

Commit:
Updated DigitizerHits histogram. New firmware looks pretty good. Need test with detector and calibration.

Commit:
Now export the discriminator time to TCagraHit. Need to export previous time too for time difference between events.

Commit:
Started to implement pole zero and asymptotic baseline corrections. Needs more work.

Commit:
Removed fake timestamp routine as we now have the myriad clock in GR

Commit:
Added fast-forward flag and functionality, which will fast-forward to the end of online raw input files.
(enabled with -F) This functionality is nice when you are in online mode (-g) and you want to skip to the end of a file to view only the newest data
as it comes.

Commit:
Added TGlobRawFile which looks in a directory for new files and attaches to them as they appears.

Commit:
Updated GRUTinizer options to utilize TGlobRawFile.

Commit:
Fully added TGlobRawFile, and removed stalled source functionality as it was preventing the sorting
of multiple CAGRA data files. This option should have a flag, so that it can be used when each file represents a unique source, but not used when
each source produces multiple files and they all need to be sorted.

Commit:
Many updates:
 - Reinstated stalled source. In the future there needs to be a distinction for online analysis between sources which produce multiple files and (per run),
   and those which produce only a single file per run. The problem was that with the stalled_source paradigm in TMultiRawFile, if a detector system produced
   a set of files sequentially (after reaching 2Gb, for example), each of those data files are only from a single physical source, but the TMultiRawFile would
   interpret them as independent physical sources, and would stall when it finished sorting the first 2Gb file (e.g.).
 - I have implemented trace analysis into TCagraHit for use with the LaBr detector in the Test Experiment. Seems to be functioning well.
 - I have added a CFD zero finder (rudimentary) for use with the 3 CFD trace pts provided per data header.

Commit:
Fully implemented PoleZero and Asymptotic Baseline corrections. I believe they are working. Need
to test against data with good peaks. Before that, I will add the flags data from the TANLEvent to TCagraHit.

Commit:
Added cal files. PoleZero & Asymptotic baseline corrections are now working. Be sure to add at least a PZ correction to the calibration file for every channel.

Commit:
Merge pull request #2 from CAGRA-GrandRaiden/cagra_corrections

Merge pull request #1 from CAGRA-GrandRaiden/cagra_corrections

Commit:
Changed MPIEventLoop submodule remote address to point to CAGRAGR organization

Commit:
Added RCNP histos file

Commit:
Added Getters for Pre/Post rise sum to TCagraHit, and added new histo loop CAGRA.cxx.
When doing polezero analysis, one should look at PreRiseSum/M vs (Post - Pre)/M to determine the polezero correction.

Commit:
Fixed bug with the saving of TChannel calibration information into the rootfile.
Added styling to mainwindow.py since it overwrites SetPadTickX/Y options applied in the logon file.
Added functions in TGrandRaiden to make it easier to plot from within the interpreter. Added new
analysis / histo file: RCNPAna.cxx.

Commit:
Error fix in last commit.

Commit:
Added fix to tcut_tab.py so that Save Cut works when the det_type is not registered. Also added many focal plane corrections.

Commit:
Cleaned repository for merge with upstream. Minimial use of defguards in main code body.
Instead, they are mostly found in source files pertaining to RCNP analysis routines which rely on externel git submodules.

Commit:
Merged upstream. Results appear identical to prior versions, however I notice a race condition when terminating the input thread.
See here for backtrace during exit: https://gist.github.com/csullivan/fd9431b459630ef0f76e70a3f961fb29

Commit:
StoppableThread was not fully merged from upstream. Fixing this resolved the race condition.

Commit:
Second pull from upstream merged. 7/19/16

Commit:
added logon script and changed rint logon script path in grutrc

Commit:
Added delay to StoppableThread::StopAll after shutting down the RCNP thread, to avoid
a deadlock owing to an incomplete push after the GRUTinizer threads have shutdown. Also changed the behavior of
the RCNP stdout so that it only prints status to screen in offline mode (ie without gui).

Commit:
Merge branch 'gr_terminate' into merge_upstream

Commit:
Removed RCNP online.bld file

Commit:
Merge pull request #3 from CAGRA-GrandRaiden/merge_upstream

Large merge from upstream (pcbend) as well as a few additions from collaboration repo

Commit:
Merge branch 'merge_upstream' into style

Commit:
Updated LaBr hits class in TGrandRaidenHits.

Commit:
Merge branch 'labr_hit_update' into merge_upstream

Commit:
Merge branch 'master' of https://github.com/pcbend/GRUTinizer into merge_upstream

Commit:
Added comment in makefile to indicate how to compile and link against RCNP libGRAnalyzer.so and associated TDetector systems.

Commit:
Removed git submodules for merging with upstream.

Commit:
Reverted README for merge with upstream.

Commit:
Merge pull request #4 from CAGRA-GrandRaiden/merge_upstream

Preparing large merge with upstream (pcbend)

csullivan commented:
I will also mention that the code was production tested during a recent DAQ test experiment in June at RCNP. This was the main reason for delaying the merge request until now (and hence why it is so large).

Lunderberg commented:
A couple comments, and requests for changes before it gets merged in.
- TCaesar.h was modified, adding the `<functional>` header.  Is this change needed?
- Name-wise, TArgonne is a bit unspecific.  Currently, each folder in TDetector is named by the detector that it unpacks.  Could it be renamed to the particular detector system being unpacked for consistency?
- The std::runtime_error being thrown by TChannel::PoleZeroCorrection.  Currently, if a value is not present, we either return a default value or nan if there is no reasonable default value. 
- What is the difference between a pedestal and a baseline?  If there is no conceptual difference, we should have them be the same value.
- The `-G` option seems to be unnecessary.  We should be able to tell whether or not the argument is a bash-style wildcard without needing to explicitly telling to program.
- Is the `-T` option still necessary?  I remember that at some point you were able to read grand raiden data from a raw binary file.  Is this still the case?
- The RCNP_BLD enum type looks to be specific to grand raiden data.  Let's change the name to reflect that.
- In the `kDetectorSystems`, `ANL` is given a value of 14.  When used in the `TUnpackingLoop::HandleGEBData`, these values and names should correspond to the official Berkeley designations for GEB-type data.  Is this value of 14 official, and what is it's official name?
- In the `kFileType` enum, there is the `ROOT_DATA` value.  This seems to be specific to Grand Raiden, not generic input data from a root file, so the name should reflect that.
- In the `TRawBanks.h`, the various `Argonne` classes should refer to the detector in question, rather than the lab.
- Could we take another look at packing the `TGrandRaiden*` into the smart buffer?  Wanting to take another look at it to see if we can avoid modifying the `TRawEvent` class with the extra `void*`.  With the calls to the grand raiden unpacker, can those be made inside `TGrandRaiden::BuildHits()`?
- The `TTreeSource` is inside the `#ifdef RCNP`.  With this being something specific to RCNP, it should have a name that reflects this.  If it is to be made more general, then it should have the ability to retrieve multiple detector branches from a single root file.  In addition, there is a hard-coded path on line 64 of `TTreeSource.h` that will break outside of nscl-land.
- Does the `TANLEvent` contain multiple individual events, each with a separate timestamp?  This is the reason why the `TGEBMode3Event`, on which it looks like the `TANLEvent` is based, is part of the event-structure rather than being part of a detector's unpacking.  If each `TANLEvent` has only a single event, then this should be moved into unpacked, rather than building.  Also, the name should be changed to indicate the detector system that uses it.
- In TGRUTint.cxx, line 152, the `if(!opt->TreeSource())` should be unnecessary.  The `OpenRootFile` will only add it to gChain if there exists a TTree named `EventTree` in it.  Is this the case for the grand-raiden-preprocessed root files?
- For the handling of the glob source in `TGRUTint::SetupPipeline`, I think I have some un-pushed code that handles it without the explicit flag.  I'll get back to you on that one.
- For the separate thread reading the Grand Raiden events from the ttree, why not just have it be an instance of `TStoppableThread`?  This would remove the necessity for the `stop_rcnp_signal` variable in `TStoppableThread.cxx`.
- In `TRawEvent::GetSize()`, is there a reason why `RCNP_BLD` returns `sizeof(void*)`?
- In `TRawEventSource::EventSource()`, anything with a `.root` filename is assumed to have RCNPEvents in it.  Is there some way we can generalize this?
- Since the target `libGRAnalyzer` does not produce a file named `libGRAnalysis`, it should be listed as a `.PHONY` target.
- In the `_load_default_style()`, we would like `style.SetOptStat()` to remain uncommented.  These are going to be moved to the `.grutrc` at some point anyways to be configurable.  Similarly, we would like for the SetPadTick commands not to be present.

csullivan commented:
Briefly skimming over things, the suggestions seem reasonable. I will address each point with a commit or comment if further discussion is needed. Thanks!

Commit:
[Pull request #110]
> TCaesar.h was modified, adding the header. Is this change needed?
Removed change to TCaeser.h that is no longer necessary due to the ifdef guard.

Commit:
[Pull request #110]
> Name-wise, TArgonne is a bit unspecific. Currently, each folder in TDetector is named by the detector that it unpacks. Could it be renamed to the particular detector system being unpacked for consistency?
I've changed it to TCagra. Note however that TANLEvent.cxx (which is generalized to different ANL detector systems) resides in the now renamed TCagra directory.

Commit:
[Pull request #110]
> The std::runtime_error being thrown by TChannel::PoleZeroCorrection. Currently, if a value is not present, we either return a default value or nan if there is no reasonable default value.
I now return a default value with a warning that is supressed after 5 consecutive calls.

Commit:
[Pull request #110]
> Is the -T option still necessary? I remember that at some point you were able to read grand raiden data from a raw binary file. Is this still the case?
I have removed TTreeSource and the corresponding flag and front end logic within TGRUTInt. As a note it still had utility
in that it could open intermediate trees, produced from the GRAnalyzer subsystem, as sources, but as this was not done generically

Commit:
[Pull request #110]
> Since the target libGRAnalyzer does not produce a file named libGRAnalysis, it should be listed as a .PHONY target.
Added as a .PHONY target.

Commit:
[Pull request #110]
> In the _load_default_style(), we would like style.SetOptStat() to remain uncommented. These are going to be moved to the .grutrc at some point anyways to be configurable. Similarly, we would like for the SetPadTick commands not to be present.
These were not meant to make it in to the pull request and have been removed.

csullivan commented:
Alright, below I respond to your comments if I have not created a commit in response to one (or multiple) of your points. If some action was needed and taken, then I have pushed a commit (see the above). 

> What is the difference between a pedestal and a baseline? If there is no conceptual difference, we should have them be the same value.

The only difference is that the Asymptotic baseline is rate dependant and therefore it’s convenient to have it be organized by timestamp. I could update the pedestal value so that it can vary with time, but this is not typically what I would expect from a “pedestal”. I’m open to suggestions.

> The -G option seems to be unnecessary. We should be able to tell whether or not the argument is a bash-style wildcard without needing to explicitly telling to program.

This flag is used for creating a TGlobRaw source, which is a wrapper on TMultiRaw that checks the globbed directories for new files that may appear after the start of a session. Unless I am misunderstanding, I think it makes sense to have this functionality be utilized via a flag. I noticed just now that in a later comment you mentioned having some unpublished source including the glob source in TGRUTInt. Perhaps this will clarify the issue.

> The RCNP_BLD enum type looks to be specific to grand raiden data. Let's change the name to reflect that.

In fact, the RCNP_BLD type is the data type for two separate spectrometers, Grand Raiden (GR), and the Large Acceptance Spectrometer (LAS). It is however a single DAQ. I can rename this to GR_BLD if you prefer, but in principle if RCNP continues to use GRUTinizer, RCNP_BLD is more consistent with their software for GR and LAS.

> In the kDetectorSystems, ANL is given a value of 14. When used in the TUnpackingLoop::HandleGEBData, these values and names should correspond to the official Berkeley designations for GEB-type data. Is this value of 14 official, and what is it's official name?

The official Berkley name is DGS for digital gammasphere, however ANL has developed the firmware of the LBL digitizers to work as general purpose digitizers. In this case, I submit that the Berkeley designation needs to be updated since ANL’s use case has evolved beyond DGS and I don’t expect them to change the firmware for each of their detector systems.

> In the TRawBanks.h, the various Argonne classes should refer to the detector in question, rather than the lab.

These raw structs are independent of a given detector system, but instead are characteristic of the utilized ANL digitizer firmware. TANLEvent is the primary unpacker class that utilizes these structs. Thus I do not believe it would be advantageous to rename these for a specific detector system.

> Could we take another look at packing the TGrandRaiden\* into the smart buffer? Wanting to take another look at it to see if we can avoid modifying the TRawEvent class with the extra void*. With the calls to the grand raiden unpacker, can those be made inside TGrandRaiden::BuildHits()?

We can do so, but a word of warning. I have spent many hours debugging a double free that occurs at the destruction of a smartbuffer (in a TRawEvent). There is a possible race condition that seems to occur. For now, I would like to leave the void\* in, until we work through this together. If I was arguing for the void*, I would say that it does allow a user to utilize the sorting components of GRUTinizer without requiring them to do the packing / unpacking. I could see this being advantageous in some cases.

> The TTreeSource is inside the #ifdef RCNP. With this being something specific to RCNP, it should have a name that reflects this. If it is to be made more general, then it should have the ability to retrieve multiple detector branches from a single root file. In addition, there is a hard-coded path on line 64 of TTreeSource.h that will break outside of nscl-land.

I have wiped this out see commit 965fd8401afdfde50c8feb975cd6744883836d0c.

> Does the TANLEvent contain multiple individual events, each with a separate timestamp? This is the reason why the TGEBMode3Event, on which it looks like the TANLEvent is based, is part of the event-structure rather than being part of a detector's unpacking. If each TANLEvent has only a single event, then this should be moved into unpacked, rather than building. Also, the name should be changed to indicate the detector system that uses it.

TANLEvent is analogous to TDDASEvent, not TGEBMode3Event. TANLEvent is a general purpose unpacking class which unpacks the GEBArgonne firmware data according to the type specified in the GEBArgonneHead struct. For this reason and those already discussed, I do not believe that TANLEvent should be modified.

> In TGRUTint.cxx, line 152, the if(!opt->TreeSource()) should be unnecessary. The OpenRootFile will only add it to gChain if there exists a TTree named EventTree in it. Is this the case for the grand-raiden-preprocessed root files?

No longer applicable, TTreeSource has been removed.

> For the handling of the glob source in TGRUTint::SetupPipeline, I think I have some un-pushed code that handles it without the explicit flag. I'll get back to you on that one.

See above discussion on TGlobRaw.

> For the separate thread reading the Grand Raiden events from the ttree, why not just have it be an instance of TStoppableThread? This would remove the necessity for the stop_rcnp_signal variable in TStoppableThread.cxx.

TTreeSource has been removed. The stop_rcnp_signal is used to stop the asynchronous thread that is running the RCNP/GR analyzer in TRCNPSource. The RCNP/GR analyzer maintains it’s own internal loop which cannot be exported out into GRUTinizer without significant changes to the RCNP code base. Since I am trying to ensure a level of maintainability between both repositories, this is the most balanced solution I have found.

> In TRawEvent::GetSize(), is there a reason why RCNP_BLD returns sizeof(void*)?

For measuring the throughput, only a pointer is being passed through the GRUTinizer sorting pipeline for each RCNP_BLD event. This is because the first level of unpacking of the RCNP BLD data is happening before it makes it into GRUTinizer. GRUTinizer is then only responsible for time correlating the RCNPEvents with the GEB events. Rather than storing the RCNPEvent class into a TSmartBuffer and then unpacking back into the same structure in TGrandRaiden (which would incur an enormous amount of copies), I simply pass a void\* pointer through GRUTinizer which can be cast out into it’s original type once it land in it’s respective TDetector system.

> In TRawEventSource::EventSource(), anything with a .root filename is assumed to have RCNPEvents in it. Is there some way we can generalize this?

This has been removed as TTreeSource is ejected.

Commit:
Merge from upstream/pcbend to resolve Pull request #110 merge conflict.

Commit:
[Pull request #110]
> Since the target libGRAnalyzer does not produce a file named libGRAnalysis, it should be listed as a .PHONY target.
In retrospect, libGRAnalyzer is a meta-rule that is marked as a dependency so that make will compile the libGRAnalyzer
submodule. Listing it as a PHONY target thus causes make to recompile all of GRUTinizer each time.

Commit:
Pole-zero and baseline vectors were missing from TChannel::Copy
which resulted in those vectors not being copied in when loading from
a rootfile.
